Added new functions: compas.datastructures.mesh.mesh_to_points

XingxinHE opened the pull request:
Dear COMPAS team, 

I hope you are doing well. I am sending my pull request.

### What type of change is this?

- [ x] New feature in `compas.datastructures.mesh.mesh_to_points`
This is a new function that samples points using the Barycentric coordinate. Yet, you can give me some suggestions like where should I put this function because I see some similar functions but without writing. Maybe they can be integrated better.

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [x ] I have added tests that prove my fix is effective or that my feature works.
- [x ] I have added necessary documentation (if appropriate)

XingxinHE committed:
initialize mesh-to-pointclouds

XingxinHE committed:
First version of mesh_to_points.py

XingxinHE committed:
Added info in AUTHORS and CHANGELOG

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+    Examples
+    --------
+
+
```
would be good to add an example

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+
+    """
+    if mesh.is_empty():
+    raise ValueError("Mesh is empty.")
```
indent

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+    if mesh.is_empty():
+    raise ValueError("Mesh is empty.")
+    if mesh.is_quadmesh():
+        mesh.quads_to_triangles()
```
if the input mesh is supposed to be a triangle mesh, you should specify this in the parameter descriptions. also, i would recommend to name the function `trimesh_to_points`.

in any case this will only convert the quads of a given mesh to triangles, but since compas supports n-gons this step is not sufficient to guarantee triangles-only...

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+    for i in range(len(mesh.face)):
+        area_list.append(mesh.face_area(i+add_idx))
+    area_list = array(area_list)
+    area_list_norm = area_list / area_list.sum()
```
this entire section can be replaced with

```
area_list = array([mesh.face_area(face) for face in mesh.faces()])
area_list_norm = area_list / area_list.sum()
```

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+
+    # (2) sample num_points on a mesh face regarding its weight of area
+    mesh_face_idx = list(mesh.face.keys())
+    sample_face_idxs = choice(mesh_face_idx, num_points, p=area_list_norm) - add_idx
```
i am assuming you created `add_idx` to make sure face identifiers map properly to array indices.
the way you are doing this now only takes care of the case when the first face identifier is not `0` assuming that from then onward the indices form a continuous and contiguous series. this is not necessarily the case. the proper way to do this would be to use a `face_index` map

```python
face_index = {face: index for index, face in enumerate(mesh.faces())}
faces = list(mesh.faces())  # unless you already did this above
sample_faces = [face_index[index] for face in choice(faces, num_points, p=area_list_norm)]
```
or 
```python
faces = list(range(mesh.number_of_faces()))
samples_faces = choice(faces, num_points, p=area_list_norm)
```

tomvanmele replied:
there is a new tutorial article related to this that might clarify things
https://compas.dev/compas/latest/tutorial/numericaldata.html

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+    samples_points = w0[ :, None] * a + w1[ :, None] * b + w2[ :, None] * c
+    return samples_points
+
+if __name__ == '__main__':
```
can't imagine this passed the linter :)

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+
+    Returns
+    -------
+    list
```
`ndarray`

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+    Parameters
+    ----------
+    mesh : compas.datastructures.Mesh
+        The mesh data structure.
```
it seems you need a triangle mesh. you should specify that here.

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+    'mesh_to_points',
+]
+
+def mesh_to_points(mesh, num_points: int = 10000):
```
i would recommend to change the name to `trimesh_samplepoints`

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+    vertices = array(vertices)
+    faces = mesh.to_vertices_and_faces()[1]
+    faces = array(faces)
+    faces_vertices = vertices[faces]
```
this can be done in one step.
for large meshes it will make a difference...

```python
# 3
vertices, faces = mesh.to_vertices_and_faces()
vertices = array(vertices, dtype=numpy.float64)
faces = array(faces)
faces_vertices = vertices[faces]
```

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+]
+
+def mesh_to_points(mesh, num_points: int = 10000):
+    """Convert a mesh to points
```
Compute sample points on a triangle mesh surface?

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+
+    # (5) Finally vertices of points
+    samples_points = w0[ :, None] * a + w1[ :, None] * b + w2[ :, None] * c
+    return samples_points
```
to summarise all the comments above :)
i think the function can be simplified to
```python
vertices, faces = mesh.to_vertices_and_faces()
vertices = array(vertices, dtype=float64)
faces = array(faces)
faces_vertices = vertices[faces]
areas = array([mesh.face_area(face) for face in mesh.faces()])
areas_normalized = areas / areas.sum()
sample_faces = choice(faces, num_points, p=areas_normalized)
# 4
...
```

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_points.py:
```diff
@@ -0,0 +1,93 @@
+    from compas.datastructures import Mesh
+
+    hypar = Mesh.from_obj(compas.get('hypar.obj'))
+    mesh = Mesh.from_obj(compas.get('faces.obj'))
```
remove or include in an actual test

tomvanmele commented on src/compas/datastructures/mesh/mesh_to_pointclouds.py:
```diff
@@ -0,0 +1,47 @@
+    'mesh_to_pointscloud',
+]
+
+def mesh_to_pointscloud(mesh, num_points: int = 10000, return_normals: bool = False):
```
i assume you are still working on an implementation?

tomvanmele reviewed (changes requested):
hi XingxinHE,

welcome to `compas-dev` :)

i have the feeling your PR is not entirely finished yet.
one of the two functions is missing an implementation and some examples and tests would be appropriate!

in the meantime already some comments about the implementation of `mesh_to_points`

XingxinHE commented:
Dear Dr. Van Mele, thanks for the reply and precious feedback. I will implement it accordingly. Have a nice weekend!

XingxinHE committed:
Revise the trimesh_samplepoints function; added new features, return normal vectors

XingxinHE commented:
@brgcode Hi Dr. Van Mele, I revised the code in light of your review.  @[10cdf90 ](https://github.com/compas-dev/compas/pull/722/commits/10cdf902369c9d873ccd394252708b6e25cb0e65)

XingxinHE commented:
I modified the file name which is consistent with the function. But it may be a bit difficult to follow up the changes from last version of code.... I am sorry about that. I summarize a bit of the changes:

(1) Example which you requested last time. I used matplotlib to visualize which might be much easier for user to understand.
```
L43 to L70
```

(2) code implementation from your review last time
```
L85 to L112
```

(3) added return normal options
```
L114 to 125
```

(4) wrap up the test script,  use `doctest.ELLIPSIS` to ignore the guid from matplotlib
```
L136
```

XingxinHE committed:
Resolve conflicts in CHANGELOG

tomvanmele commented on src/compas/datastructures/mesh/trimesh_samplepoints.py:
```diff
@@ -0,0 +1,136 @@
+        raise ValueError("Mesh is invalid.")
+
+    # (1)  acquire the area of all mesh faces
+    area_list = array([mesh.face_area(face) for face in mesh.faces()])
```
FYI, depending on the size of your mesh, it might be faster to do this numerically as well
```python
V = array(vertices)
F = array(faces, dtype=int)

e01 = V[F[:, 1]] - V[F[:, 0]]
e12 = V[F[:, 2]] - V[F[:, 1]]
e20 = V[F[:, 0]] - V[F[:, 2]]

normal = np.cross(e01, e12)
A2 = normrow(normal)
A = 0.5 * A2.ravel()
```

tomvanmele replied:
as an example, see `compas.numerical.geodesics_numpy`

tomvanmele commented on src/compas/datastructures/mesh/trimesh_samplepoints.py:
```diff
@@ -0,0 +1,136 @@
+    area_list_norm = area_list / area_list.sum()
+
+    # (2) sample num_points on a mesh face regarding its weight of area
+    faces = list(range(mesh.number_of_faces()))
```
i would rename this to `nf` or `f` since it creates confusion wrt to the actual faces...

tomvanmele commented on src/compas/datastructures/mesh/trimesh_samplepoints.py:
```diff
@@ -0,0 +1,136 @@
+    # (3) create a ndarray of vertices regarding the faces
+    vertices, faces = mesh.to_vertices_and_faces()
+    vertices = array(vertices, dtype=float64)
+    faces = array(faces)
```
perhaps this can be done in the beginning and be used for the calculation of the areas as well (see above)

tomvanmele commented on src/compas/datastructures/mesh/trimesh_samplepoints.py:
```diff
@@ -0,0 +1,136 @@
+
+        return samples_points, samples_points_normals
+    else:
+        return samples_points
```
there is no need for the `else`, since you return from the preceding `if`

tomvanmele commented on src/compas/datastructures/mesh/trimesh_samplepoints.py:
```diff
@@ -0,0 +1,136 @@
+]
+
+
+def trimesh_samplepoints(mesh, num_points: int = 1000, return_normals: bool = False):
```
the convention in compas is that functions using `numpy/scipy` are suffixed with `_numpy`. this helps clarifying which functions are available in Rhino/GH directly and which are not...

tomvanmele replied:
this is also true for the module name

tomvanmele replied:
also, since you are providing type hints (and this function will not be loaded in Rhino/GH directly anyway), perhaps consider adding the return type hint as well
```python
def trimesh_samplepoints_numpy(mesh: compas.datastructures.Mesh, num_points: int = 1000, return_normals: bool = False) -> Union[np.ndarray, Tuple[np.ndarray, np.ndarray]]:
```

tomvanmele commented on src/compas/datastructures/mesh/trimesh_samplepoints.py:
```diff
@@ -0,0 +1,136 @@
+    (if True) samples_points_normals(numpy.ndarray)
+        A numpy array representing the normal vector of sampled points  with dim = [num_points, 3]
+
+    Examples
```
this is a bit hard to digest. consider splitting it up in a few steps with comments
```rst
Examples
--------
Make a triangle mesh.
>>> ...

Compute sample points.
>>> ...

Visualize with Matplotlib.
>>> ...

tomvanmele reviewed (commented):
generally looking good! just a few suggestions...
only thing missing is the import to bring it up to the API level.
in `compas.datastructures.mesh.__init__.py`
```python
if not compas.IPY:
    from .trimesh_samplepoints_numpy import *
```
and an entry in list of mesh functions of the datastructures package.
see the docstring of `compas.datastructures.__init__.py`

XingxinHE committed:
Revise trimesh_samplepoints_numpy

XingxinHE committed:
Rename the .py file

XingxinHE committed:
Update the CHANGLOG

XingxinHE commented:
Hi Dr. Van Mele@brgcode , thank you for the advice and for taking the time to explain the convention which is super helpful for my future contribution. OK, here are my commits. 
@[d17074d](https://github.com/compas-dev/compas/pull/722/commits/d17074d701531acb069b9e70987fb9a02ad6f21a) This is everything I changed excluding the name of .py file which may be easier for you to catch up on what I modified. 
@[0c7853a](https://github.com/compas-dev/compas/pull/722/commits/0c7853a8f38d21b85f3a3320e52cc26b23105519) Modify the name of .py file. 
@[7a5d302](https://github.com/compas-dev/compas/pull/722/commits/7a5d302ce233f3c751ea4ecca63117ed6c05e57a) I forgot to modify the name on CHANGELOG... oops... 

That is it! Thanks!

tomvanmele committed:
Merge branch 'master' into mesh-to-pointclouds

tomvanmele committed:
remove cyclic import

tomvanmele committed:
fixed typing

tomvanmele committed:
Update trimesh_samplepoints_numpy.py
